### PR TITLE
MTDSA-13247: API #1413 List Payments And Allocation Details Alignment Updates

### PR DESCRIPTION
**Pull request by tom-lloyd-hmrc**
- [Spec](https://confluence.tools.tax.service.gov.uk/pages/viewpage.action?pageId=323192033)
- [Ticket](https://jira.tools.tax.service.gov.uk/browse/MTDSA-13247) 

Along with the changes in the spec, I noticed the Payment Lot Item regex was incorrect.

In some places, the V1 endpoints had the same error code but different error messages. For these, I've prefixed the error object with `V1_` and then created a `V2_` version with the updated message.

### CODE REVIEW

File: app/v2/controllers/requestParsers/validators/ListPaymentsAndAllocationDetailsValidator.scala [hunk 1]
```diff
@@ -17,27 +17,36 @@
...
   private def parameterValidation: ListPaymentsAndAllocationDetailsRawData => List[List[MtdError]] =
     (data: ListPaymentsAndAllocationDetailsRawData) => {
       List(
         NinoValidation.validate(data.nino),
-        StartEndDateValidation.validate(data.dateFrom, data.dateTo),
-        PaymentLotValidation.validate(data.paymentLot),
+        DateFormatValidation.validate(data.fromDate, V2_FromDateFormatError),
+        DateFormatValidation.validate(data.toDate, V2_ToDateFormatError),
+        PaymentLotValidation.validateFormat(data.paymentLot),
         PaymentLotItemValidation.validate(data.paymentLotItem)
```
Comment (jeremystone) [inline app/v2/controllers/requestParsers/validators/ListPaymentsAndAllocationDetailsValidator.scala:38]: Minor but given the link between payment lot and payment lot item, that one method indicates that it validates format and the other doesn't could be made more obvious. Maybe rename to `PaymentLotItemValidation.validateFormat`. 

Alternatively (more work but arguably nicer) given that `validateMissing` was (arbitrarily?) added to `PaymentLotValidation` rather than `PaymentLotItemValidation` you could ditch `PaymentLotItemValidation` module altogether and move its methods to `PaymentLotValidation`, with `validateMissingPaymentLot`, `validatePaymentLotFormat` and `validatePaymentLotItemFormat` methods.

File: test/api/controllers/requestParsers/validators/validations/DateRangeValidationV2Spec.scala [hunk 2]
```diff
@@ -33,15 +33,22 @@ class DateRangeValidationV2Spec extends UnitSpec {
         DateRangeValidationV2.validate(None, None) shouldBe List()
       }
     }
-    "return a list containing an error" when {
+
+    "return a list containing a RangeToDateBeforeFromDateError" when {
```
Comment (jeremystone) [inline test/api/controllers/requestParsers/validators/validations/DateRangeValidationV2Spec.scala:37]: Not officially part of the change I realise but this should also check the boundary conditions:
- it's ok when to and from represent the same date
- it's an error when to date represents the day before the from date

**Review by jeremystone — COMMENTED** [review]
Just a few minor things.

File: app/api/models/errors/MtdErrors.scala [hunk 3]
```diff
@@ -46,24 +46,36 @@ object TaxYearFormatError
...
+object V1_FromDateFormatError
     extends MtdError(
       code = "FORMAT_FROM_DATE",
       message = "The provided From date is invalid"
     )
 
+object V2_FromDateFormatError
+    extends MtdError(
+      code = "FORMAT_FROM_DATE",
+      message = "The provided fromDate is invalid."
```
[the remark this answers is not in the document]
Comment (tom-lloyd-hmrc) [reply app/api/models/errors/MtdErrors.scala]: Didn't realise that, updated!

File: app/v2/controllers/requestParsers/validators/ListPaymentsAndAllocationDetailsValidator.scala (hunk 1 above)
Comment (tom-lloyd-hmrc) [reply app/v2/controllers/requestParsers/validators/ListPaymentsAndAllocationDetailsValidator.scala:38]: I added the `validateMissing` method to `PaymentLotValidation` because the error code is specifically about the `paymentLot` field, not the `paymentLotItem` field. I think I'll go with the rename of the validate function, and keep the two validations if that's ok.
